cachi2/core/package_managers/gomod.py

```diff
@@ -582,7 +582,37 @@ def _resolve_gomod(
         # Make Go ignore the vendor dir even if there is one
         go_list.extend(["-mod", "readonly"])
 
-    main_module_name = _run_gomod_cmd([*go_list, "-m"], run_params).rstrip()
+    go_list_modules = _run_gomod_cmd([*go_list, "-m", "-json"], run_params).rstrip()
+    module_list = list(load_json_stream(go_list_modules))
+    raw_main_module = module_list.pop(0)
+    main_module_name = raw_main_module["Path"]
+
+    def get_relative_path_workspace_to_main_module(path1: Path, path2: Path) -> str:
+        """
+        Return relative path from workspace module to main module.
+        """
+        if not path2.is_relative_to(app_dir.root):
+            raise Exception(
+                f"Workspace module path {path2} is not relative to root path {app_dir.root}."
+            )
+        relative_path = os.path.relpath(path2, path1)
+        return relative_path
+
+    workspace_modules = []
+    if len(module_list) > 0:
+        for module in module_list:
+            workspace_modules.append(
+                ParsedModule(
+                    path=module["Path"],
+                    version=version_resolver.get_golang_version(main_module_name, app_dir),
+                    replace=ParsedModule(
+                        path=get_relative_path_workspace_to_main_module(
+                            Path(raw_main_module["Dir"]), Path(module["Dir"])
+                        )
```

Comment:
I'd instantiate the replacement ParsedModule separately in a variable to enhance readability a bit.